### PR TITLE
fix: Fix assembled response in prompt logs #447

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/util/MergeChunks.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import lombok.experimental.UtilityClass;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+@UtilityClass
+public class MergeChunks {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/util/MergeChunks.java:19]: please add original reference (link to the python file)

File: src/main/java/com/epam/aidial/core/util/MergeChunks.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            return source;
+        } else if (target.getNodeType() == JsonNodeType.STRING && source.getNodeType() == JsonNodeType.STRING) {
+            String text = target.textValue() + source.textValue();
+            return new TextNode(text);
+        }
+
+        throw new IllegalArgumentException(String.format("Can't merge %s into %s at path %s", source.asText(), target.asText(), toPath(path)));
+    }
+
+    private static String toPath(Stack<String> path) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/util/MergeChunks.java]: Nit: return String.join(".", path);

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java [hunk 3]
```diff
@@ -297,34 +271,77 @@ static String assembleStreamingResponse(Buffer response) {
...
             if (statistics != null) {
+                MergeChunks.removeIndices(statistics);
                 result.set("statistics", statistics);
             }
             if (systemFingerprint != null) {
                 result.set("system_fingerprint", systemFingerprint);
             }
 
-            if (content.isEmpty()) {
+            if (choices == null) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/log/GfLogStore.java]: let's make this code same as above
if (choices != null) {
remove indices and set
}

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java [hunk 4]
```diff
@@ -297,34 +271,77 @@ static String assembleStreamingResponse(Buffer response) {
...
-            message.put("content", content.toString());
-
             return ProxyUtil.convertToString(result);
         } catch (Throwable e) {
             log.warn("Can't assemble streaming response", e);
             return "{}";
         }
     }
 
+    private static void mergeCustomContent(ObjectNode merged, ObjectNode cur) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/log/GfLogStore.java]: remove dead code

File: src/main/java/com/epam/aidial/core/util/MergeChunks.java (hunk 1 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/util/MergeChunks.java:19]: fixed

File: src/main/java/com/epam/aidial/core/util/MergeChunks.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/util/MergeChunks.java]: fixed

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/log/GfLogStore.java]: fixed

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java (hunk 4 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/log/GfLogStore.java]: fixed

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java [hunk 5]
```diff
@@ -297,27 +271,24 @@ static String assembleStreamingResponse(Buffer response) {
...
             }
             if (systemFingerprint != null) {
                 result.set("system_fingerprint", systemFingerprint);
             }
 
-            if (content.isEmpty()) {
+            if (choices != null) {
+                MergeChunks.removeIndices(choices);
+                result.set("choices", choices);
+            } else {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/log/GfLogStore.java]: Nit: else block is redundant here, the same code as below

**Review by Maxim-Gadalov — DISMISSED** [review]

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java (hunk 5 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/log/GfLogStore.java]: right fixed.

**Review by Maxim-Gadalov — APPROVED** [review]

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-448.nightly-test.deltixhub.io
>E2E tests status: failed

**Pull request merged**